snuba/query/parser/__init__.py

```diff
@@ -200,3 +216,181 @@ def apply_aliases(exp: Expression) -> Expression:
             return replace(exp, alias=exp.column_name)
 
     query.transform_expressions(apply_aliases)
+
+
+def _expand_aliases(query: Query) -> None:
+    """
+    Recursively expand all the references to aliases in the query. This
+    makes life easy to query processors and translators that only have to
+    take care of not introducing shadowing (easy to enforce at runtime),
+    otherwise aliasing is transparent to them.
+    """
+    # Pre-inline all nested aliases. This reduces the number of iterations
+    # we need to do on the query.
+    # Example:
+    # {"a": f(x), "x": g(k)} -> {"a": f(g(k)), "x": g(k)}
```

Comment:
This is just an optimization to do fewer scans over the query that does not compromise readability to me. When there is a nested expression that defines an alias and contains references to other aliases, if we pre-expand that expression enttirely, when we expand the references in the query, the expression is already fully expanded and we do not need to expand it again each time.